include/vkfw/vkfw.hpp

```
@@ -3139,7 +3139,11 @@ namespace VKFW_NAMESPACE {
 
 #ifndef VKFW_NO_SMART_HANDLE
 template <> struct VKFW_NAMESPACE::CustomDestroy<VKFW_NAMESPACE::Instance> {
-  VKFW_INLINE void destroy(Instance &instance) { static_cast<void>(instance.destroy()); }
+  VKFW_INLINE void destroy(Instance &instance) {
+    try {
+      static_cast<void>(instance.destroy());
+    } catch (...) {}
```

I'd wrap `try` and `catch` into an `#ifndef VKFW_HPP_NO_EXCEPTIONS`.

Also, I'd add another warning (`pragma message`) inside the catch block, so a warning about discarding errors is printed for anyone who would compile destructors with exception support.

---

So the function should look like this?
```c++
#ifndef VKFW_HPP_NO_EXCEPTIONS
static_cast<void>(instance.destroy());
#else
try {
   static_cast<void>(instance.destry());
} catch (...) {}
#endif
```

As for the warning/message: having that message there will add noise to the compiler output that isn't really necessary as the behaviour is already documented in the `README`. In addition, the message would have to be enabled independently of whether exceptions are enabled because the current behaviour already swallows error return codes in the event exceptions are disabled.
If you really believe that another warning is necessary, then there also needs to be a flag to silence it. That would mean the destroy functions would look something like this:
```c++
  VKFW_INLINE void destroy(Instance &instance) {
#ifndef VKFW_NO_HANDLE_DESTRUCTOR_WARNING
#pragma message("warning: unique handle destructors will swallow any exceptions or return codes. You can silence this warning by defining VKFW_NO_HANDLE_DESTRUCTOR_WARNING")
#endif
#ifndef VKFW_HPP_NO_EXCEPTIONS
   static_cast<void>(instance.destroy());
#else
    try {
      static_cast<void>(instance.destroy());
    } catch (...) {}
#endif
```

Thoughts?

---

To follow up: having the compiler issue a warning message is a bad idea in this case because *all* users would get it unless they explicitly disabled the handles. So if a user forgets to disable the handles (or doesn't want to for whatever reason) they now get that warning message on their builds whether they actually use the handles or not! 
Even having the flag to disable it would be annoying, because I'd have to silence a warning for something that I may not even use.

---

Yes, I realize that this will produce a warning by default. But don't forget that we're _discarding errors_ by default. Which is what worries me the most.

But I also don't want to disable automatic handles by default either - because they are like _the_ feature of these bindings.

In my mind, the way to go is something like
```cpp
// (untested)
UniqueHandle::~UniqueHandle() VKFW_NOEXCEPT {
#ifndef VKFW_HPP_NO_EXCEPTIONS
  try {
#endif

    if (m_value)
      this->destroy(m_value);

#ifndef VKFW_HPP_NO_EXCEPTIONS
  } catch(...) {
    // DANGER!!!

    #ifndef VKFW_ALLOW_DESTRUCTORS_TO_DISCARD_EXCEPTIONS
      #pragma message("warning: unique handle destructors will NEVER throw. "
                      "Which means exceptions passing through them WILL BE lost. "
                      "You can either disable exception altogether (by defining "
                      "`VKFW_HPP_NO_EXCEPTIONS`) or silence this warning (by defining "
                      "`VKFW_ALLOW_DESTRUCTORS_TO_DISCARD_EXCEPTIONS`)."
                      "If you use the second approach it's highly recommended you handle such "
                      "errors in another way, for example using `vkfw::setErrorCallback`");"
    #endif
  }
#endif
}
and then each user (who cares about warnings) to decide which way is better for them.

Of course it would also mean fully disabling the usage of `vk::UniqueHandle` in favor of `vkfw::UniqueHandle`, which is sad, but not a deal-breaker I'd say.

--------------------------------------------

But honestly, as I'm typing this out, I'm realizing that this might be an overkill. Considering even `vulkan_hpp` doesn't seem to do anything special to handle this corner case, and that's a huge and well maintained library 🤔.

---

BTW, it looks like there's a bit of a misunderstanding. All the different flavors of `destroy` we have can stay the way they are. It's not like they are prevented from throwing! It's specifically destructors (as in, just `~UniqueHandle`) that is the problem.

---

I honestly believe that the warning is completely overkill and too heavy handed for something like this. I can change the note on the `README` to a warning so it's more emphatic, but having it as a compiler warning that will appear whether you use the handles or not is excessive.

> BTW, it looks like there's a bit of a misunderstanding. All the different flavors of destroy we have can stay the way they are. It's not like they are prevented from throwing! It's specifically destructors (as in, just ~UniqueHandle) that is the problem.

I'm not changing all of them! The snippet I pasted might've been a bit misleading though. The only `destroy` functions I'm changing are the ones that belong to the `vkfw::CustomDestroy` structs. That way, it maintains compatibility with both `vkfw::UniqueHandle` and `vk::UniqueHandle`. Even then, I'm specifically *only* changing the ones for `vkfw::Instance` and `vkfw::Window`, since `vkfw::Cursor` can't throw anyway.

---

> I honestly believe that the warning is completely overkill and too heavy handed for something like this

Honestly, sure, let's turn the compilation-time warning into a comment instead. If someone runs into trouble with it, they'll probably be able to find it anyway.

> the ones that belong to the `vkfw::CustomDestroy` structs

Ah, ok, then misunderstanding is all mine! Sorry for the noise (it's been a while since I touched this code :D)

> since `vkfw::Cursor` can't throw anyway.

Wait, it can raise errors, specifically [https://www.glfw.org/docs/latest/group__input.html#ga81b952cd1764274d0db7fb3c5a79ba6a:~:text=Errors-,Possible%20errors%20include%20GLFW_NOT_INITIALIZED%20and%20GLFW_PLATFORM_ERROR.,-Reentrancy], meaning it can throw!

> I should probably change the wording of the README to reflect the fact that only `vkfw::UniqueInstance` and `vkfw::UniqueWindow` do this.

Yes, please!

---

> Wait, it can raise errors, specifically [GLFW_NOT_INITIALIZED and GLFW_PLATFORM_ERROR](https://www.glfw.org/docs/latest/group__input.html#ga81b952cd1764274d0db7fb3c5a79ba6a:~:text=Errors-,Possible%20errors%20include%20GLFW_NOT_INITIALIZED%20and%20GLFW_PLATFORM_ERROR.,-Reentrancy), meaning it can throw!

> Practically, think destroying the instance before destroying the cursor object.

Not according to the implementation of `vkfw::Cursor::destroy()`. That will call `createResultValue` with `Result::VKFW_ENUMERATOR(Success)`. which is unlike the `destroy` for the window/instance where those call `getError()`. 

Unless that was a mistake and it was meant to call `getError()` instead of always returning success?

---

Yes, it sounds like a bug. Could you, please, take care of it too while at it?